packages/wrangler/src/api/startDevWorker/DevEnv.ts

```diff
@@ -93,6 +95,34 @@ export class DevEnv extends EventEmitter {
 		});
 	}
 
+	/**
+	 * Awaits for all the local runtime controllers to be ready, and when they are
+	 * it resolves a promise on the proxy controller to let it know that the
+	 * local server is now ready to handle requests
+	 */
+	private async resolveAllRuntimeControllersReady(): Promise<void> {
+		await Promise.all(
+			this.runtimes
+				.filter(
+					(runtime) =>
+						runtime instanceof RemoteRuntimeController ||
+						runtime instanceof LocalRuntimeController
+				)
+				.map((runtime) => {
+					return new Promise<void>((resolve) => {
+						// A runtime controller is ready when it either emits the
+						// reloadComplete event (meaning that the runtime is in use and
+						// ready) or when it emits the teardown event (meaning that the
+						// runtime is not currently in use)
+						runtime.once("reloadComplete", () => resolve());
+						runtime.once("teardown", () => resolve());
+					});
```

Comment:
Should this be a promise race?

Reply:
no, we want all the runtime controllers to be ready

for example the runtime controller's teardown even will fire much earlier than the local runtime controller that could be building a docker image, and the whole point of this change is to show the ready-on log after the docker build is completed